### PR TITLE
Implement enumerator for functions.

### PR DESCRIPTION
**Pull request by ajreynol**
This will allow extensionality in higher-order QF_UF to be applied in a lazier fashion.

### CODE REVIEW

**Comment by ajreynol** [comment]
This fixes #1177

File: src/theory/builtin/theory_builtin_rewriter.cpp [hunk 1]
```diff
@@ -99,6 +99,29 @@ RewriteResponse TheoryBuiltinRewriter::postRewrite(TNode node) {
   }
 }
 
+TypeNode TheoryBuiltinRewriter::getFunctionTypeForArrayType( TypeNode atn, Node bvl ) {
+  std::vector<TypeNode> children;
+  for( unsigned i=0; i<bvl.getNumChildren(); i++ ){
+    Assert( atn.isArray() );
+    Assert( bvl[i].getType()==atn.getArrayIndexType() );
```
Comment (timothy-king) [inline src/theory/builtin/theory_builtin_rewriter.cpp]: Can this fail in practice, i.e. the count for the atn is different than bvl.getNumChildren()?

File: src/theory/builtin/theory_builtin_rewriter.cpp [hunk 2]
```diff
@@ -99,6 +99,29 @@ RewriteResponse TheoryBuiltinRewriter::postRewrite(TNode node) {
...
+  children.push_back( atn );
+  return NodeManager::currentNM()->mkFunctionType( children );
+}
+
+TypeNode TheoryBuiltinRewriter::getArrayTypeForFunctionType( TypeNode ftn ) {
+  Assert( ftn.isFunction() );
+  // construct the curried array type
+  unsigned nchildren = ftn.getNumChildren();
+  TypeNode ret = ftn[nchildren-1];
+  for( int i=((int)nchildren-2); i>=0; i-- ){
```
Comment (timothy-king) [inline src/theory/builtin/theory_builtin_rewriter.cpp]: static_cast

File: src/theory/builtin/type_enumerator.h [hunk 3]
```diff
@@ -23,6 +23,7 @@
 #include "expr/type_node.h"
 #include "expr/uninterpreted_constant.h"
 #include "theory/type_enumerator.h"
+#include "theory/builtin/theory_builtin_rewriter.h"
```
Comment (timothy-king) [inline src/theory/builtin/type_enumerator.h]: Not sure this is the right order. clang-format on this region?

File: src/theory/builtin/type_enumerator.h [hunk 4]
```diff
@@ -75,6 +76,49 @@ class UninterpretedSortEnumerator : public TypeEnumeratorBase<UninterpretedSortE
...
 };/* class UninterpretedSortEnumerator */
 
+
+/** FunctionEnumerator 
+* This enumerates function values, based on the enumerator for the
+* array type corresponding to the given function type.
+*/
+class FunctionEnumerator : public TypeEnumeratorBase<FunctionEnumerator> {
+public:
+  FunctionEnumerator(TypeNode type, TypeEnumeratorProperties * tep = NULL) :
```
Comment (timothy-king) [inline src/theory/builtin/type_enumerator.h]: nullptr

Comment (timothy-king) [inline src/theory/builtin/type_enumerator.h]: move into cpp

File: src/theory/builtin/type_enumerator.h [hunk 5]
```diff
@@ -75,6 +76,49 @@ class UninterpretedSortEnumerator : public TypeEnumeratorBase<UninterpretedSortE
...
+  /** get the current term of the enumerator */
+  Node operator*() {
+    if(isFinished()) {
+      throw NoMoreValuesException(getType());
+    }
+    Node a = *d_arrayEnum;
+    return TheoryBuiltinRewriter::getLambdaForArrayRepresentation( a, d_bvl );
+  }
+  /** increment the enumerator */
+  FunctionEnumerator& operator++() throw() {
```
Comment (timothy-king) [inline src/theory/builtin/type_enumerator.h]: Add override

File: src/theory/builtin/type_enumerator.h [hunk 6]
```diff
@@ -75,6 +76,49 @@ class UninterpretedSortEnumerator : public TypeEnumeratorBase<UninterpretedSortE
...
+    ++d_arrayEnum;
+    return *this;
+  }
+  /** is the enumerator finished? */
+  bool isFinished() throw() {
```
Comment (timothy-king) [inline src/theory/builtin/type_enumerator.h]: add override

File: src/theory/builtin/type_enumerator.h [hunk 7]
```diff
@@ -75,6 +76,49 @@ class UninterpretedSortEnumerator : public TypeEnumeratorBase<UninterpretedSortE
...
+    return d_arrayEnum.isFinished();
+  }
+private:
+  /** type properties */
+  TypeEnumeratorProperties * d_tep;
```
Comment (timothy-king) [inline src/theory/builtin/type_enumerator.h]: This appears to be unused. Remove.

File: src/theory/builtin/type_enumerator.h (hunk 5 above)
Comment (timothy-king) [reply src/theory/builtin/type_enumerator.h]: Move into cpp

File: src/theory/builtin/type_enumerator.h [hunk 8]
```diff
@@ -75,6 +76,49 @@ class UninterpretedSortEnumerator : public TypeEnumeratorBase<UninterpretedSortE
 
 };/* class UninterpretedSortEnumerator */
 
+
```
Comment (timothy-king) [inline src/theory/builtin/type_enumerator.h]: Give this its own header & c++ file

File: src/theory/builtin/theory_builtin_rewriter.h [hunk 9]
```diff
@@ -62,6 +62,16 @@ class TheoryBuiltinRewriter {
   /** recursive helper for getArrayRepresentationForLambda */
   static Node getArrayRepresentationForLambdaRec( TNode n, bool reqConst, TypeNode retType );
 public:
+  /** Get function type for array type
+  * This returns the function type of terms returned by the function getLambdaForArrayRepresentation( t, bvl ),
+  * where t.getType()=atn.
+  */
+  static TypeNode getFunctionTypeForArrayType( TypeNode atn, Node bvl );
```
Comment (timothy-king) [inline src/theory/builtin/theory_builtin_rewriter.h]: Why are these public and static in TheoryBuiltinRewriter? Who needs to access them? Do these need special access to anything in TheoryBuiltinRewriter?

File: src/theory/builtin/type_enumerator.h [hunk 10]
```diff
@@ -75,6 +76,49 @@ class UninterpretedSortEnumerator : public TypeEnumeratorBase<UninterpretedSortE
...
+class FunctionEnumerator : public TypeEnumeratorBase<FunctionEnumerator> {
+public:
+  FunctionEnumerator(TypeNode type, TypeEnumeratorProperties * tep = NULL) :
+    TypeEnumeratorBase<FunctionEnumerator>(type),
+    d_arrayEnum(TheoryBuiltinRewriter::getArrayTypeForFunctionType(type), tep) {
+    Assert(type.getKind() == kind::FUNCTION_TYPE);
+    d_bvl = NodeManager::currentNM()->getBoundVarListForFunctionType( type );
+  }
+  /** get the current term of the enumerator */
+  Node operator*() {
```
Comment (timothy-king) [inline src/theory/builtin/type_enumerator.h]: Add override.
Move into cpp.

File: src/theory/builtin/type_enumerator.h (hunk 4 above)
Comment (timothy-king) [inline src/theory/builtin/type_enumerator.h]: What is the memory lifetime of tep expected to be?

File: src/theory/builtin/theory_builtin_rewriter.cpp (hunk 1 above)
Comment (ajreynol) [reply src/theory/builtin/theory_builtin_rewriter.cpp]: Yes, it is up to the caller to ensure that they give an appropriate bvl/atn pair.

File: src/theory/builtin/theory_builtin_rewriter.h (hunk 9 above)
Comment (ajreynol) [reply src/theory/builtin/theory_builtin_rewriter.h]: They are useful in the conversion between functions and arrays.  FunctionEnumerator is an example of something that needs them.

File: src/theory/builtin/type_enumerator.h (hunk 4 above)
Comment (ajreynol) [reply src/theory/builtin/type_enumerator.h]: tep is not owned by the TypeEnumerator.  I've added some comments in theory/type_enumerator.h to make this clear.

File: src/theory/builtin/type_enumerator.h (hunk 8 above)
Comment (ajreynol) [reply src/theory/builtin/type_enumerator.h]: I think this is overkill.  The precedent we've had is to group type rules / type enumerators into a single file per theory.

**Comment by ajreynol** [comment]
I've addressed the review in my latest commit.

File: src/theory/builtin/type_enumerator.h [hunk 11]
```diff
@@ -75,6 +76,31 @@ class UninterpretedSortEnumerator : public TypeEnumeratorBase<UninterpretedSortE
...
+*/
+class FunctionEnumerator : public TypeEnumeratorBase<FunctionEnumerator> {
+public:
+  FunctionEnumerator(TypeNode type, TypeEnumeratorProperties * tep = nullptr);
+  /** Get the current term of the enumerator. */
+  Node operator*() override;
+  /** Increment the enumerator. */
+  FunctionEnumerator& operator++() throw() override;
+  /** is the enumerator finished? */
+  bool isFinished() throw() {
```
Comment (timothy-king) [inline src/theory/builtin/type_enumerator.h]: Add override to isFinished.

File: src/theory/builtin/type_enumerator.h [hunk 12]
```diff
@@ -75,6 +76,31 @@ class UninterpretedSortEnumerator : public TypeEnumeratorBase<UninterpretedSortE
...
+* array type corresponding to the given function type.
+*/
+class FunctionEnumerator : public TypeEnumeratorBase<FunctionEnumerator> {
+public:
+  FunctionEnumerator(TypeNode type, TypeEnumeratorProperties * tep = nullptr);
+  /** Get the current term of the enumerator. */
+  Node operator*() override;
+  /** Increment the enumerator. */
+  FunctionEnumerator& operator++() throw() override;
+  /** is the enumerator finished? */
```
Comment (timothy-king) [inline src/theory/builtin/type_enumerator.h]: I am not sure the comments on the overriden functions are informative beyond their contract on TypeEnumeratorInterface. I think they can safely be dropped.

File: src/theory/builtin/type_enumerator.h (hunk 8 above)
Comment (timothy-king) [reply src/theory/builtin/type_enumerator.h]: Not sure I agree, but adding the cpp is already a big step forward so lets proceed.

File: src/theory/builtin/theory_builtin_rewriter.h (hunk 9 above)
Comment (timothy-king) [reply src/theory/builtin/theory_builtin_rewriter.h]: From your description it sounds like these are unrelated to TheoryBuiltinRewriter then. Consider making these free functions.

File: src/theory/builtin/type_enumerator.h (hunk 12 above)
Comment (ajreynol) [reply src/theory/builtin/type_enumerator.h]: I am following the policy that seems to be in place that all functions in h files need comments.  Let me know if this is not the case and if there is a style guideline for this.

File: src/theory/builtin/theory_builtin_rewriter.h (hunk 9 above)
Comment (ajreynol) [reply src/theory/builtin/theory_builtin_rewriter.h]: These functions should be in the same place as getLambdaForArrayRepresentation and getArrayRepresentationForLambda, which are used in the builtin rewriter.  At the moment I think this is the best place for these 4 functions.

**Comment by ajreynol** [comment]
I've addressed the review in my latest commit.

**Comment by ajreynol** [comment]
This fixes bug reported on #1345, I'm promoting this to major.

File: src/theory/builtin/theory_builtin_rewriter.h [hunk 13]
```diff
@@ -62,31 +62,58 @@ class TheoryBuiltinRewriter {
...
+ * the index types of the
+ * (curried) Array type. For example a bound variable list bvl whose variables
+ * have types (Int, Real) would be a proper
+ * input when paired with atn = (Array Int (Array Real Bool)), or (Array Int
+ * (Array Real (Array Bool Bool))).
+ * This function returns (-> Int Real Bool) and (-> Int Real (Array Bool Bool))
+ * respectively in these cases.
+ * Whereas bvl is not a proper input for atn = (Array Int (Array Bool Bool)) or
+ * (Array Int Int).
+ */
```
Comment (aniemetz) [inline src/theory/builtin/theory_builtin_rewriter.h]: Fix formatting (hard to read).

(I know, this is mainly due to the fact that clang-format simply breaks comment lines that exceed the 80 char limit, but as of yet there's no way to let clang-format handle that in a better way. Unfortunately this means that we will sometimes have to fix this manually.)

**Comment by ajreynol** [comment]
I've reformatted the comment and improved it in my latest commit.

**Pull request merged**